### PR TITLE
Cleanup of Rubi

### PR DESCRIPTION
**Pull request by oscargus**
#### Brief description of what is fixed or changed
Based on LGTM I decided to cleanup parts of the Rubi code. Primarily by removing unused imports etc. This is a rather tedious work, so I push an intermediate version to get some feedback, especially if it is a stupid idea for some reason to remove all unused imports etc.

#### Other comments
As I didn't really start out with Rubi, there are some odd other cleanup as well.

Will be interesting to see if this has a notable impact on the import time of Rubi and if that may lead to that tests can be enabled.

Finally, I am reserving the right to simply get bored with this and not cleaning up everything in one go.

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by Upabjojr** [comment]
i see you are editing some auto-generated files. Did you update the code generator as well? You may risk that the next time the code is regenerated, they unused imports will reappear.

> Primarily by removing unused imports etc. This is a rather tedious work, [...]

What about automated tools?

**Comment by oscargus** [comment]
> i see you are editing some auto-generated files. Did you update the code generator as well? You may risk that the next time the code is regenerated, the unused imports will reappear.
Ah, thanks for letting me know! I was, sort of, worried that it might be the case.

> > Primarily by removing unused imports etc. This is a rather tedious work, [...]
> 
> What about automated tools?
Yes, that would indeed be useful here. Clearly, there are tools that find them (otherwise I would not have found them...), but I'll try to find something that removes it as well (and ideally add to the generator).

**Comment by oscarbenjamin** [comment]
@oscargus should this be closed?

**Comment by oscargus** [comment]
I have not had the time/inspiration to look at it in a while...

The main problem is that some of the changes will be overwritten the next time the rubi code is automatically regenerated. Hence, a bit of the work in is vain. Still, some of it is not, so I think that a decent solution would be to merge it (after resolving the conflicts) and live with that some of the changes will be reverted eventually. The option is that I remove changes in the files that will be regenerated, but that seems to be more work for no apparent gain.

**Comment by oscarbenjamin** [comment]
That seems fine to me. +1 to merge when tests pass

**Comment by Upabjojr** [comment]
Please don't edit the generated code. I would rather suggest to edit the code generator in order to produce clean code.

Next time someone runs the code generator all issues will return.

**Comment by oscargus** [comment]
@Upabjojr I completely agree (now that I know it is generated). Here, the current question is really: a).merge as is to get the other non-generated stuff in and live with that some of the edits were in vain or b) revert the edits in generated files by hand before merging. I'd vote for a).

**Comment by Upabjojr** [comment]
I was actually reviewing the code generator yesterday. I believe it is necessary to split the generated files in two parts.

On the formatting side, what about adding an auto formatter to the code generator instead? The is a folder called `parsetools` where with the code generators, maybe a call to the code formatter may be added there?

This is a huge PR in terms of bytes permanently added to the SymPy repo.

**Comment by oscargus** [comment]
> This is a huge PR in terms of bytes permanently added to the SymPy repo.

Yes, that is correct. I guess I can just restore the automatically generated files. (For some reason I thought it was complicated, more or less had to do it by hand, but that is of course not the case.)

**Comment by oscargus** [comment]
https://pypi.org/project/autoflake/ may be useful to remove unused imports in the generated files automatically. 

https://pypi.org/project/autopep8/ could probably be used for formatting.

**Comment by oscargus** [comment]
I spent some time playing with automatic cleanups and it seems like autoflake cannot handle imports within ().

One possible solution is to use https://github.com/timothycrosley/isort to split imports into one import per line, run autoflake, run isort to reformat imports. However, it does not handle imports within if-clauses.

So, right now what I have is a bit of line breaking from autopep8 and some rare import removals from autoflakes plus some removed unused variables (however, the computation still happens, which is sometimes the expected behaviour, sometimes not). This can be slightly improved by reformatting at least parts of the import template to not use () (some lines plainly do not have to use it).

This is my current test script

```
import sys
from autopep8 import main as autopep8_main
from autoflake import _main as autoflake_main
from isort import SortImports

#Changed order to start with a smaller file...
files =['piecewise_linear.py', 'hyperbolic.py', 'integrand_simplification.py', 'linear_products.py', 'quadratic_products.py', 'binomial_products.py', 'trinomial_products.py',
        'miscellaneous_algebraic.py', 'exponential.py', 'logarithms.py', 'sine.py', 'tangent.py', 'secant.py', 'miscellaneous_trig.py',
        'inverse_trig.py',  'inverse_hyperbolic.py', 'special_functions.py', 'miscellaneous_integration.py']

for file in files:
    args = ["autopep8", file, "-i", "-a"]
    print(file)
    autopep8_main(args)
    arguments={}
    arguments['force_single_line'] = True
    SortImports(file, **arguments)
    args = ["autoflake", file, "-i", "--remove-all-unused-imports", "--remove-unused-variables"]
    autoflake_main(args, standard_out=sys.stdout, standard_error=sys.stderr)
    arguments['multi_line_output'] = 4
    SortImports(file, **arguments)
```

**Comment by Upabjojr** [comment]
There is a template file called `header.py.txt`, or something similar. The imports could be turned into a template field, (this may make things easier).

Anyway, I suggest to ignore RUBI for now. That module is still not in a usable state and needs some more urgent work than code cleanup.

File: sympy/integrals/rubi/tests/test_utility_function.py [hunk 1]
```diff
@@ -456,19 +568,18 @@ def test_FractionalPowerFreeQ():
     assert FractionalPowerFreeQ(x)
 
 def test_Exponent():
-    assert Exponent(x**2 + x + 1 + 5, x, Min) == 0
```
Comment (smichr) [inline sympy/integrals/rubi/tests/test_utility_function.py]: Why were these tests changed? Especially if they were passing before they look like they are testing something different now.

Comment (oscargus) [reply sympy/integrals/rubi/tests/test_utility_function.py]: `Exponent` used to have a third optional argument which is a function that is applied to the result. This is in practice only used in `utility_function.py` for converting to a list. LGTM complained about the function sometimes returning a scalar argument, sometimes an iterable. I decided to split `Exponent` into `Exponent` and `ExponentList`. (Only `Exponent` is used in the generated code as far as I can tell, `ExponentList` is exclusively used in `utility_function`.)

Hence, testing that `Exponent` can accept a `Min` argument seemed like not really required...

**Comment by czgdp1807** [comment]
From the last few comments, it looks like that this PR can be closed. However, I am not sure. Please let me know, what should we do to conclude this. Thanks for the contributions.
ping @Upabjojr

**Comment by Upabjojr** [comment]
Actually this PR looks pretty good. I have re-checked and no auto-generated files are being edited, which is good.

Only care... most RUBI tests are disabled by default on Travis, someone should check that they all pass.

**Comment by Upabjojr** [comment]
If tests pass, I'm +1 to merge this PR.

**Comment by czgdp1807** [comment]
Pushed the latest commits. If tests pass, I will merge it.

**Pull request merged**